rr_iarrc/src/finish_line_watcher/finish_line_watcher.cpp

```diff
@@ -78,23 +74,85 @@ void ImageCB(const sensor_msgs::ImageConstPtr& msg) {
     frame = cv_ptr->image;
     blockEnvironment(frame);
 
-    auto count = countNonZero(frame);
+    // Do contour detection
+    vector<vector<Point>> contours;
+    vector<Vec4i> hierarchy;
+    Mat contourImg = frame.clone();
+    findContours(contourImg, contours, hierarchy, CV_RETR_LIST, CV_CHAIN_APPROX_SIMPLE);
+
+    Mat drawing = Mat::zeros(contourImg.size(), CV_8UC3);
+    double maxLength = 0.0;
+    double maxLengthYPos = 0.0;
+    for (size_t i = 0; i < contours.size(); i++) {
+        // Find the average slopes of the contour
+        Point start = contours.at(i).at(0);
+        Point end = contours.at(i).at((contours.at(i).size() - 1) / 2);
+        double rise = start.y - end.y;
+        double run = start.x - end.x + 0.01;
+        double slope = std::abs(rise / run);
+        // Also find the length
+        double length = arcLength(contours.at(i), false);
+
+        // Filter by slope and contour
+        if (slope > 0.0001 && slope < slope_cutoff && length > length_cutoff) {
+            // Draw for debug image
+            drawContours(drawing, contours, (int)i, Scalar(255, 0, 0), 2, LINE_8, hierarchy, 0);
+
+            if (length > maxLength) {
+                maxLength = length;
+                maxLengthYPos = contours.at(i).at((int)(contours.at(i).size() / 2)).y;
+            }
+
+        } else {
+            drawContours(drawing, contours, (int)i, Scalar(0, 255, 0), 2, LINE_8, hierarchy, 0);
+        }
+    }
 
-    auto width = getWidth(frame);
+    bool detected = maxLength > length_cutoff;
 
-    if (state == LOW && count > 2000 && width > 700) {
-        state = HIGH;
-    } else if (state == HIGH && count < 2000) {
-        state = LOW;
-        number_of_crosses++;
-        ROS_INFO_STREAM("Finish line crossed: " << to_string(number_of_crosses));
+    if (publish_when_detected) {
```

Comment:
Here, we want the functionality to always uptick the number of crosses. The difference is when it ups the value, either as soon as it sees a valid finish line or as soon as we lose sight of a seen finish line.

Reply:
I (think) that's what this code is doing; if we are upticking when the line is first detected (`publish_when_detected=true`) then state will be low and, if the cooldown is at 0 and the line is detected, `if (detected && state == LOW && cooldown == 0)` will be true so the value will be increased and then the state is reset by the following else if. If we are upticking when we lose sight of the line then the opposite happens: the state to high when the line is detected then when we lose sight of the line then we reset the state to low and increase the value.